### PR TITLE
26 create subscription resource

### LINKED ISSUE #26: Create subscription resource in FHIR server once follow up request is received
This will enable the requesting system to be notified once a follow up request has been done.

### PR DESCRIPTION
**Pull request by njogz**
# Description

This PR creates a subscription resource in the FHIR server which makes it easier to inform the requesting service once a follow up task has been completed. It also calls the CHT's API endpoint for submitting a record so as to create a task

Closes #26 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.
 
# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: mediator/src/controllers/service-request.ts [hunk 1]
```diff
@@ -19,18 +22,41 @@ async function createServiceRequest(request: ServiceRequest) {
...
-    const res = await axios.get(`${fhirUrl}/Patient/${patientId}`, options);
+    const res = await axios.get(`${fhirUrl}/Patient/?identifier=${patientId}`, options);
 
     if (res.status !== 200) {
       return {status: res.status, data: res.data};
     }
 
-    // todo => @njogz to add task creation process.
+    // generate subscription resource
+    const FHIRSubscriptionResource = generateFHIRSubscriptionResource(patientId, callbackURL);
```
Comment (samuelimoisili) [inline mediator/src/controllers/service-request.ts:32]: We had a discussion about intercepting the FHIR subscription at Zanzibar. Are we still implementing that?

File: mediator/src/controllers/service-request.ts [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+import { generateFHIRSubscriptionResource } from '../utils/subscription';
+import { generateApiUrl } from '../utils/service-request';
 
-const axios = require('axios');
-const {FHIR} = require('../../config');
 const {url: fhirUrl, password: fhirPassword, username: fhirUsername} = FHIR;
 
 type ServiceRequest = {
   patient_id: string;
+  callbackURL: string;
```
[the remark this answers is not in the document]
Comment (samuelimoisili) [reply mediator/src/controllers/service-request.ts:11]: Yeah, that's for variable names in JavaScript, but this is JSON request. The `app_settings.json` uses snake_case for variable naming in JSON.

File: mediator/src/controllers/service-request.ts (hunk 1 above)
Comment (samuelimoisili) [reply mediator/src/controllers/service-request.ts:32]: Oh, okay. Should I create an issue for that?

File: mediator/src/controllers/service-request.ts [hunk 3]
```diff
@@ -19,18 +22,41 @@ async function createServiceRequest(request: ServiceRequest) {
...
+    });
+
+    if (chtRes.data.success !== true) {
+      // TODO: delete the subscription
+      return {status: 500, data: 'unable to create the follow up task'};
+    }
+
+    logger.info(JSON.stringify(chtRes.data, null, 4));
+
+    return {status: res.status};
```
[the remark this answers is not in the document]
Comment (samuelimoisili) [reply mediator/src/controllers/service-request.ts:55]: Yeah, that works. Status code gets the job done, but sending a subscription resource to the requesting system will be beneficial in the future if we want the requesting system to track the progress of the task, cancel callbacks, update callback url (maybe the domain was migrated) or cancel the task in the future.

[the remark this answers is not in the document]
Comment (samuelimoisili) [reply mediator/src/controllers/service-request.ts:55]: We can discuss about this in the proposed design meeting.

**Review by samuelimoisili — APPROVED** [review]

**Review by andrablaj — COMMENTED** [review]
@njogz would it be possible to enumerate the steps and give examples of payloads to test this PR? I would like to reproduce the task creation in the CHT. Thank you!

File: mediator/src/middlewares/schemas/service-request.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import joi from 'joi';
 
 export const createServiceSchema = joi.object({
-  patient_id: joi.number().required()
+  patient_id: joi.number().required(),
```
Comment (andrablaj) [inline mediator/src/middlewares/schemas/service-request.ts:4]: @njogz, when posting the patient/callbackURL payload to service-request, as advised in [your step-by-step guide](https://github.com/medic/interoperability/issues/26#issuecomment-1441451470),  I get:

>     "message": "\"patient_id\" must be a number",

and that is normal, as the schema required a number and not a uuid. anything that I am missing here?

**Review by andrablaj — CHANGES_REQUESTED** [review]
Sharing an error I bumped in while running the [step-by-step guide](https://github.com/medic/interoperability/issues/26#issuecomment-1441451470) you suggested.

File: README.md [hunk 5]
```diff
@@ -30,6 +30,8 @@ A first version of the project can be found in the [chis-interoperability](https
...
 1. Ensure you have [cht-conf](https://www.npmjs.com/package/cht-conf) installed and run `cht --local` to compile and upload the app settings configuration to your local CHT instance.## Shutdown the servers
+1. Create a new user in the CHT instance with the username `interop-client` using these [instructions](https://docs.communityhealthtoolkit.org/apps/tutorials/contact-and-users-1/#4-create-the-chw-user). For the role you can select `Data entry` and `Analytics` roles. Please note that you can use any username you prefer but you would have to update the the config with the new username. You can do that by editing the `cht-config/app_settings.json` file and updating the `username` value in the `outbound` object e.g. on this [line](https://github.com/medic/interoperability/blob/main/cht-config/app_settings.json#L452).
+1. Securely save the `interop-client` user's password to the database using the instructions [here](https://docs.communityhealthtoolkit.org/apps/reference/api/#credentials). Change the values `mykey` and `my pass` to `openhim1` and your user's password respectively.
```
Comment (andrablaj) [inline README.md:34]: I would add an example of `PUT` request, just to make it extra clear:
`curl -X PUT -H "Content-Type: text/plain" http://medic:password@localhost:5988/api/v1/credentials/openhim1 -d 'interop-password'`

File: README.md [hunk 6]
```diff
@@ -30,6 +30,8 @@ A first version of the project can be found in the [chis-interoperability](https
 1. Set up a local CHT instance using theses [instructions](https://docs.communityhealthtoolkit.org/apps/tutorials/local-setup/)
 1. Set up a proxy to your local CHT instance by running using something like [nginx-local-ip](https://github.com/medic/nginx-local-ip) or [ngrok](https://ngrok.com/) and update the `CHT_URL` value in the `.env` file with the new URL.
```
Comment (andrablaj) [inline README.md:31]: > value in the `.env` file with the new URL.

I suggest adding extra step before this step and mentioning the `.env` file creation, and the existence of the `env.template` file.
Mentioning `CHT_USERNAME` & `CHT_PASSWORD` can be helpful too, as the users might use different credentials.

File: mediator/src/controllers/service-request.ts [hunk 7]
```diff
@@ -1,16 +1,19 @@
+import axios from 'axios';
 import {logger} from '../../logger';
+import {FHIR, CHT} from '../../config';
```
Comment (andrablaj) [inline mediator/src/controllers/service-request.ts:3]: There are some formatting differences, and I assume they would be fixed via #43 ?

File: mediator/src/controllers/service-request.ts [hunk 8]
```diff
@@ -19,18 +22,40 @@ async function createServiceRequest(request: ServiceRequest) {
...
+    const chtApiUrl = generateApiUrl(CHT.url, CHT.username, CHT.password);
+    const chtRes = await axios.post(chtApiUrl, {
+      _meta: {
+        form: "interop_follow_up",
+      },
+      patient_uuid: patientId,
+    });
+
+    if (chtRes.data.success !== true) {
+      // TODO: delete the subscription
```
Comment (andrablaj) [inline mediator/src/controllers/service-request.ts:49]: Do we have an issue to tackle this `TODO`? If not, I recommend opening one (+ explaining why we need to delete the subscription). There is no such step in the sequence diagram, and it's worth a discussion on wether we need to delete it or not.

**Review by andrablaj — CHANGES_REQUESTED** [review]
@njogz so happy that I managed to reproduce all the steps in the flow locally, great job on this issue!

My comments are mostly around documentation and style. 

While running the flow locally, I noticed that the logs were not clear enough, I imagine one step to solve this would be #40.

I also noticed the lack of unit testing, and I an not sure if you were considering them via another issue.

**Review by andrablaj — APPROVED** [review]
This looks good, thank you for the documentation updates @njogz

**Pull request merged**